### PR TITLE
write SplitMulti in python (take 2)

### LINKED ISSUE #4023: port split_multi to Python in the style of filter_alleles

### LINKED ISSUE #4024: support split_multi on tables
Depends on: https://github.com/hail-is/hail/issues/4023

### PR DESCRIPTION
**Pull request by catoverdrive**
I think this is finally at a place where it's not a whole lot slower than the current version. I timed this with two changes that are not yet in master: #4069, and a change that adds a shuffle flag to MatrixMapRows (to shuffle instead of coerce if we're changing the key). I haven't broken that out as a separate change yet because I'm not super sure about it.

I've put some timings below (`profile` and `profile75` are left aligned; `chr22` and `profile225` contain moving variants).

```
profile    |  new, no shuffle  | [ 38.77559779,  19.34779241,  30.00672880]
profile    |  old              | [ 20.66201923,  22.53258884,  19.24919411]

profile75  |  old              | [ 57.11803599,  55.57265945,  55.78256949]
profile75  |  new, no shuffle  | [ 84.90397896,  47.73250838,  48.00151558]
profile75  |  new, shuffle     | [102.61078402,  52.47020745,  47.33059412]

profile225 |  old              | [166.27919411, 169.35251976, 162.12195974]
profile225 |  new, no shuffle  | [573.15814854, 249.25716389, 249.55162573]
profile225 |  new, shuffle     | [288.81559407, 130.46181546, 129.05702945]

chr22      |  old              | [343.85546785, 344.91024148, 337.22106485]
chr22      |  new, no shuffle  | [916.14408915, 497.77703920, 497.592671299]
```
cc @cseed 

edit: updated timings with the fix for RepartitionedRDD:

```
profile225-cut      new      108.9945367 053.5403599 050.8885137
profile225          new      249.1994185 134.2645221 132.4985398
chr22               new      269.2498286 266.3412299 265.5198199
```

(profile225-cut is the same size as profile75, but with a non-left-aligned variant)

(also: fixes #4023, fixes #4024)

### CODE REVIEW

**Comment by catoverdrive** [comment]
update: `calculateKeyRanges` doesn't work correctly when there's only one element (which we never hit, since `coerce` considers those to be sorted), which was causing the shuffle to drop the only moved element in some of my tests (which is why the tests were running so much faster than the non-shuffling version). 

I did fix something in the ReorderedPartitions dependencies that was causing an extra partition to be included in the dependencies during the `union_rows`, and now this branch (sans explode_rows changes) is running at about the same speed as SplitMulti was previously, although there's still generally noticeable bump for the first iteration.

so @tpoterba I think this is probably set for review.

File: python/hail/methods/statgen.py [hunk 1]
```diff
@@ -1309,9 +1310,32 @@ def pc_relate(call_expr, min_individual_maf, *, k=None, scores_expr=None,
...
 
-class SplitMulti(object):
+
+@typecheck(ds=oneof(Table, MatrixTable),
+           keep_star=bool,
+           left_aligned=bool)
+def split_multi(ds, keep_star=False, left_aligned=False) -> Union[Table, MatrixTable]:
     """Split multiallelic variants.
 
+    The resulting dataset will be keyed by the split locus and alleles.
```
Comment (catoverdrive) [inline python/hail/methods/statgen.py]: These docs are mostly (almost entirely) copied from the `SplitMulti` class and the `split_multi_hts` docs.

**Comment by konradjk** [comment]
Did I never PR the VEP splitting logic? I don't see it, so I must not have. I'll wait for this to go in and then do it then

**Comment by tpoterba** [comment]
discuss post and merge at will

**Comment by tpoterba** [comment]
I merged at will

**Pull request merged**